### ISSUE TITLE
Prometheus staleness markers are generated for metrics that aren't stale.

### ISSUE THREAD

**Issue by a-feld** [open-telemetry/opentelemetry-collector-contrib#4748]
**Describe the bug**

Staleness markers may be incorrectly generated for replicated services (jobs with multiple targets).

_Setup_
* 2 replicated prometheus endpoints (same metric set, except for service_instance_id)
* Example configuration:
```yaml
global:
  scrape_interval: 5s
  scrape_timeout: 5s
  evaluation_interval: 5s
scrape_configs:
  - job_name: "kubernetes-pods"
    static_configs:
      - targets: ["localhost:8000", "localhost:8001"]
```

**Steps to reproduce**
* Deploy 2 instances of the opentelemetry-collector.
* Configure the prometheus receiver to scrape both instances.
* Configure the logging exporter in deubg mode

Observe metric values with NaN are reported.

**What did you expect to see?**
No NaN values should be reported.

**What did you see instead?**
NaN metric values were reported.

**What version did you use?**
Version: 0.30.1

**What config did you use?**
```yaml
receivers:
  prometheus:
    config:
      global:
        scrape_interval: 5s
        scrape_timeout: 5s
        evaluation_interval: 5s
      scrape_configs:
        - job_name: "kubernetes-pods"
          static_configs:
            - targets: ["localhost:8000", "localhost:8001"]

exporters:
  logging:
    logLevel: debug

service:
  pipelines:
    metrics:
      receivers: [prometheus]
      processors: []
      exporters: [logging]
```

**Environment**
OS: MacOS and hyperkit (docker/k8s)
Compiler(if manually compiled): go1.16

**Additional context**
I believe what could be happening is the labelset / metric names are the same for both replicas, with the exception of **service_instance_id**

* Each pod is scraped independently
* Since the first replica’s service_instance_id label is not observed when the second replica is scraped, I believe this is causing NaNs to be emitted for the first pod’s scrape (to be confirmed?), deleting the state - even though the metrics aren’t actually stale

**Comment by odeke-em** [comment]
I can reproduce this failure with this test that end to end implements the described situation
```go
package repro

import (
	"context"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"

	"github.com/gogo/protobuf/proto"
	"github.com/golang/snappy"
	"github.com/prometheus/prometheus/pkg/value"
	"github.com/prometheus/prometheus/prompb"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"go.opentelemetry.io/collector/component"
	"go.opentelemetry.io/collector/processor/batchprocessor"
	"go.opentelemetry.io/collector/service"
	"go.opentelemetry.io/collector/service/parserprovider"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"

	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter"
	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver"
)

// Test that staleness markers are NOT emitted for replicated services whose scrapes might alternate
// between intervals.
// Ideally after this test, which runs the entire collector and end-to-end scrapes then checks with the
// Prometheus remotewrite exporter should count NO staleness markers emitted.
// See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/4748
func TestStalenessMarkersReplicatedServicesDoNotProduceStalenessMarkers(t *testing.T) {
	if testing.Short() {
		t.Skip("This test can take a long time")
	}

	ctx, cancel := context.WithCancel(context.Background())

	createStalenessHandler := func() http.HandlerFunc {
		// 1. Setup the server.
		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
			select {
			case <-ctx.Done():
				return
			default:
			}

			fmt.Fprintf(rw, `
# HELP jvm_memory_bytes_used Used bytes of a given JVM memory area.
# TYPE jvm_memory_bytes_used gauge
jvm_memory_bytes_used{area="heap"} %.1f`, float64(i))
		})
	}

	scrapeServer1 := httptest.NewServer(createStalenessHandler())
	defer scrapeServer1.Close()
	scrapeServer2 := httptest.NewServer(createStalenessHandler())
	defer scrapeServer2.Close()

	serverURL1, err := url.Parse(scrapeServer1.URL)
	require.Nil(t, err)
	serverURL2, err := url.Parse(scrapeServer2.URL)
	require.Nil(t, err)

	// 2. Set up the Prometheus RemoteWrite endpoint.
	prweUploads := make(chan *prompb.WriteRequest)
	prweServer := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
		// Snappy decode the uploads.
		payload, rerr := ioutil.ReadAll(req.Body)
		if err != nil {
			panic(rerr)
		}
		recv := make([]byte, len(payload))
		decoded, derr := snappy.Decode(recv, payload)
		if err != nil {
			panic(derr)
		}

		writeReq := new(prompb.WriteRequest)
		if uerr := proto.Unmarshal(decoded, writeReq); uerr != nil {
			panic(uerr)
		}

		select {
		case <-ctx.Done():
			return
		case prweUploads <- writeReq:
		}
	}))
	defer prweServer.Close()

	// 3. Set the OpenTelemetry Prometheus receiver.
	config := fmt.Sprintf(`
receivers:
  prometheus:
    config:
      scrape_configs:
        - job_name: 'test'
          scrape_interval: 2ms
          static_configs:
            - targets: [%q,%q]

processors:
  batch:
exporters:
  prometheusremotewrite:
    endpoint: %q
    insecure: true

service:
  pipelines:
    metrics:
      receivers: [prometheus]
      processors: [batch]
      exporters: [prometheusremotewrite]`, serverURL1.Host, serverURL2.Host, prweServer.URL)

	// 4. Run the OpenTelemetry Collector.
	receivers, err := component.MakeReceiverFactoryMap(prometheusreceiver.NewFactory())
	require.Nil(t, err)
	exporters, err := component.MakeExporterFactoryMap(prometheusremotewriteexporter.NewFactory())
	require.Nil(t, err)
	processors, err := component.MakeProcessorFactoryMap(batchprocessor.NewFactory())
	require.Nil(t, err)

	factories := component.Factories{
		Receivers:  receivers,
		Exporters:  exporters,
		Processors: processors,
	}

	appSettings := service.CollectorSettings{
		Factories:      factories,
		ParserProvider: parserprovider.NewInMemory(strings.NewReader(config)),
		BuildInfo: component.BuildInfo{
			Command:     "otelcol",
			Description: "OpenTelemetry Collector",
			Version:     "tests",
		},
		LoggingOptions: []zap.Option{
			// Turn off the verbose logging from the collector.
			zap.WrapCore(func(zapcore.Core) zapcore.Core {
				return zapcore.NewNopCore()
			}),
		},
	}
	app, err := service.New(appSettings)
	require.Nil(t, err)

	go func() {
		if err := app.Run(); err != nil {
			t.Error(err)
		}
	}()

	// Wait until the collector has actually started.
	stateChannel := app.GetStateChannel()
	for notYetStarted := true; notYetStarted; {
		switch state := <-stateChannel; state {
		case service.Running, service.Closed, service.Closing:
			notYetStarted = false
		}
	}

	// The OpenTelemetry collector has a data race because it closes
	// a channel while
	if false {
		defer app.Shutdown()
	}

	// 5. Let's wait on 10 fetches.
	var wReqL []*prompb.WriteRequest
	for i := 0; i < 10; i++ {
		wReqL = append(wReqL, <-prweUploads)
	}
	defer cancel()

	// 6. Assert that we encounter the stale markers aka special NaNs for the various time series.
	staleMarkerCount := 0
	totalSamples := 0
	for i, wReq := range wReqL {
		name := fmt.Sprintf("WriteRequest#%d", i)
		require.True(t, len(wReq.Timeseries) > 0, "Expecting at least 1 timeSeries for:: "+name)
		for j, ts := range wReq.Timeseries {
			fullName := fmt.Sprintf("%s/TimeSeries#%d", name, j)
			assert.True(t, len(ts.Samples) > 0, "Expected at least 1 Sample in:: "+fullName)

			// We are strictly counting series directly included in the scrapes, and no
			// internal timeseries like "up" nor "scrape_seconds" etc.
			metricName := ""
			for _, label := range ts.Labels {
				if label.Name == "__name__" {
					metricName = label.Value
				}
			}
			if !strings.HasPrefix(metricName, "jvm") {
				continue
			}

			for _, sample := range ts.Samples {
				totalSamples++
				if value.IsStaleNaN(sample.Value) {
					staleMarkerCount++
				}
			}
		}
	}

	require.True(t, totalSamples > 0, "Expected at least 1 sample")
	require.False(t, staleMarkerCount > 0, fmt.Sprintf("Expected no stale markers, got: %d", staleMarkerCount))
}
```

**Comment by codeboten** [comment]
Was this issue addressed by #5062?

**Comment by Aneurysm9** [comment]
> Was this issue addressed by #5062?

Yes, thi should have been addressed there.  @a-feld please reopen if this is incorrect.
